File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    
    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    static boolean[] visited;
    static int answer = 0;
    static StringBuilder sb = new StringBuilder();

    
    
    public int solution(int n, int[][] computers) {
        
        // [1] 그래프 세팅 
        visited = new boolean[n];
        
        for(int i=0; i<n; i++){
            graph.add(new ArrayList<>());
        }
        
        
        
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if (i == j) continue;
                
                if (computers[i][j] == 1){
                    graph.get(i).add(j);   
                }
                
            }
        }
        
        
        
        for(int i =0; i<n; i++){
            System.out.println(graph.get(i));
        }
        
        
        
        
        // [2] dfs 실행
        for(int i = 0; i<n; i++){
            if(!visited[i]){
                
                dfs(i);
                answer++;
                
                sb.append("\n");
                    
            }
        }
        
        System.out.println(sb.toString());
        
        return answer;

    
    }
    
    
    static void dfs(int v){
        visited[v] = true;
        sb.append(v).append(" ");
        
        
        for(int next: graph.get(v)){
            if(!visited[next]){
                dfs(next);
            }
            
        }
        
        
        
        
        
        
        
    }
    
    
    
    
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import defaultdict, deque


def solution(n, computers):
    graph = defaultdict(set)
    visited = [0 for i in range(n)]
    
    cnt = 0
    
    for i in range(n):
        for j in range(n):
            if i == j:
                continue
            if computers[i][j] == 1:
                graph[i].add(j)
                graph[j].add(i)

                
                
                
    def bfs(start):
        queue = deque()
        queue.append(start)
        visited[start] = 1
        
        while queue:
            v = queue.popleft()
            
            for i in graph[v]:
                if not visited[i]:
                    visited[i] = 1
                    queue.append(i)
                    
    
    
    for i in range(n):
        if not visited[i]:
            bfs(i)
            cnt += 1
    

    
    return cnt
        
        
        
        
        